Implement persistent locks and bans

sirDonovan opened the pull request:
Write locks and bans to a .csv file to be loaded upon server
start/restart

Slayer95 commented on users.js:
```diff
@@ -95,6 +96,48 @@ var lockedUsers = Object.create(null);
+	var buffer = '';
+	for (var ip in bannedIps) {
+		if (!bannedIps[ip].time) continue;
+		buffer += "ban|" + bannedIps[ip].time + "|" + bannedIps[ip].userid + "|" + ip + "|ip\n";
```
Maybe make the first value <code>banip</code> rather than splitting the "ip" part as a separate value?
(Same for locks)

Slayer95 commented:
Why pipe-delimited values? We should either use JSON or stick to CSV as other <code>users</code> features do.

Zarel commented:
I'd prefer TSV in the future; CSV ends up being sort of dumb in my experience.

Zarel commented on commands.js:
```diff
@@ -1481,6 +1481,7 @@ var commands = exports.commands = {
 			return;
 		}
 		room.destroyLog(function () {
+			Users.exportLocksBans();
```
So crashes would reset locks/bans?

Joimer replied:
Yeah, unless you back them up setting a timer.

Joimer commented:
Donovan, better use json for server-side objects (check how rooms are stored). Also, remember our conversation? Perhaps add a timer to back bans and locks up once every... 12 hours? What do you think, @Zarel?

Zarel commented:
I don't like JSON. We use it for rooms only because rooms are too complicated for TSV/CSV.

On second thought, let's go with CSV.

How often we backup bans/locks depends on how slow the backup process is.

sirDonovan commented:
I've updated the code to delete expired locks and bans each time the file is loaded and also added the backup interval, currently set to 12 hours. It writes a 110KB file (total of 1000 entries) in 4ms (for reference, we've had about 350 locks and bans in the past 3 days). 

I can't easily replicate the server's load so I don't know how big of a difference there is between that and running it locally, but do you think the backup should run more/less frequently?

Zarel commented:
12 hours is fine for a 4ms runtime.

Zarel commented on commands.js:
```diff
@@ -1481,6 +1481,8 @@ var commands = exports.commands = {
 			return;
 		}
 		room.destroyLog(function () {
+			clearInterval(Users.backupLocksBans);
```
no

(I can't think of a way `/kill` could fail, but if it does, PS should clearly keep backing up locks and bans)

Zarel commented on users.js:
```diff
@@ -95,6 +96,56 @@ var lockedUsers = Object.create(null);
+}
+Users.importLocksBans = importLocksBans;
+Users.exportLocksBans = exportLocksBans;
+Users.backupLocksBans = setInterval(function () {
```
end a variable name with `Timer` if it's a timer. :|

Zarel commented on users.js:
```diff
@@ -1231,41 +1300,47 @@ User = (function () {
+		var entry = {
+			time: time,
+			userid: userid
+		};
```
Having entries being structured like this increases RAM usage and slows down access by a bit.

With 300ish active, it shouldn't be a big problem, though. We _do_ periodically clean them up, right?

Zarel replied:
I think I would personally use an array, though. `[time, userid]`.

Honestly, both arrays and objects take up way too much RAM in JavaScript. They're not very lightweight. :/

Zarel commented on users.js:
```diff
@@ -95,6 +96,56 @@ var lockedUsers = Object.create(null);
+	var buffer = '';
+	for (var ip in bannedIps) {
+		if (!bannedIps[ip].time) continue;
+		buffer += "banip," + bannedIps[ip].time + "," + bannedIps[ip].userid + "," + ip + "\n";
```
Apparently we _don't_ periodically clean them up.

We should.

You know how when importing, we don't import bans/locks that have already expired?

We shouldn't export bans/locks that have already expired. In fact, we should delete them while exporting.

Zarel commented on users.js:
```diff
@@ -473,7 +524,15 @@ User = (function () {
+		this.locked = false;
+		var checkResult = Users.checkLocked(connection.ip);
+		if (checkResult) {
+			if (checkResult.time && Date.now() - checkResult.time > PERSISTENT_PUNISHMENT_TIME) {
```
Shouldn't this sort of thing happen _inside_ `Users.checkLocked`?

Zarel commented on users.js:
```diff
@@ -1231,41 +1300,47 @@ User = (function () {
+	User.prototype.ban = function (noRecurse, userid, time) {
 		// recurse only once; the root for-loop already bans everything with your IP
 		if (!userid) userid = this.userid;
+		if (!time) time = Date.now();
```
Why would start time matter? Wouldn't expiration time make more sense?

Zarel replied:
I guess both would be useful to track. But expiration time far more so.

Zarel commented on users.js:
```diff
@@ -1273,10 +1348,17 @@ User = (function () {
+			userid: this.userid
+		};
+		if (this.autoconfirmed) lockedUsers[this.autoconfirmed] = entry;
+		if (this.authenticated) lockedUsers[this.userid] = entry;
```
Using the pointer to the same entry? How risky is this?

I suppose the entry is basically constant, so not too much.

sirDonovan commented:
I've updated the code.

Zarel commented on app.js:
```diff
@@ -391,7 +391,9 @@ Rooms.global.formatListText = Rooms.global.getFormatListText();
+// load locks and bans at our leisure
+Users.importLocksBans();
+
 fs.readFile('./config/ipbans.txt', function (err, data) {
```
Shouldn't we move this inside `importLocksBans`, too?

sirDonovan replied:
Oh, yeah that makes sense. Updated again.

Zarel commented on users.js:
```diff
@@ -26,6 +26,7 @@
 const THROTTLE_DELAY = 600;
 const THROTTLE_BUFFER_LIMIT = 6;
 const THROTTLE_MULTILINE_WARN = 4;
+const PERSISTENT_PUNISHMENT_TIME = 3 * 24 * 60 * 60 * 1000;
```
Let's split this into LOCK_TIME 36 hours and BAN_TIME 3 days

Zarel commented on users.js:
```diff
@@ -95,6 +96,86 @@ var lockedUsers = Object.create(null);
 var rangelockedUsers = Object.create(null);
 
+function importLocksBans() {
+	fs.readFile('config/locksbans.csv', function (err, data) {
```
I'm still not sure how I think about `locksbans.csv` as the name. Maybe something like `punishments.csv`.

Zarel replied:
`sanctions.csv`?

Zarel commented on users.js:
```diff
@@ -95,6 +96,86 @@ var lockedUsers = Object.create(null);
+			if (row[0] === 'banip') {
+				bannedIps[row[3]] = entry;
+			} else if (row[0] === 'ban') {
+				bannedUsers[row[3]] = entry;
```
Wait, banning IPs and usernames are done on separate rows? Why not all on one row?

sirDonovan replied:
I'm not sure what you mean. When the file is exported, the first values on each line (<code>row[0]</code> here) differentiate locked/banned IPs and usernames because they are stored in different objects.

Zarel replied:
I've thought a bit more about this and I guess it's fine for now.

I still think the 'banip' and 'ban' lines can be combined.

Zarel commented on users.js:
```diff
@@ -95,6 +96,86 @@ var lockedUsers = Object.create(null);
+Users.exportLocksBans = exportLocksBans;
+Users.backupLocksBansTimer = setInterval(function () {
+	exportLocksBans();
+}, 12 * 60 * 60 * 1000);
```
You know you can just do

    setInterval(exportLocksBans, 12 * 60 * 60 * 1000);

...right?

Zarel commented on users.js:
```diff
@@ -112,10 +193,28 @@ function ipSearch(ip, table) {
-	return ipSearch(ip, bannedIps);
+	var checkResult = ipSearch(ip, bannedIps);
+	if (!checkResult) return false;
+	if (Array.isArray(checkResult)) {
```
Is it possible for the result not to be an array?

Zarel replied:
Considering how you use it elsewhere, it appears the answer is "no".

Zarel replied:
Or your other code is very unsafe.

sirDonovan replied:
The result can be <code>'#cflood'</code> if the IP is banned by the ResourceMonitor and <code>'#ipban'</code> if the IP is banned with /banip or in the ipbans file

Zarel commented on users.js:
```diff
@@ -1231,41 +1340,44 @@ User = (function () {
+	User.prototype.ban = function (noRecurse, userid, expiration) {
 		// recurse only once; the root for-loop already bans everything with your IP
 		if (!userid) userid = this.userid;
+		if (!expiration) expiration = PERSISTENT_PUNISHMENT_TIME + Date.now();
```
Should we allow negative numbers to mean "never expire"?

Zarel commented on users.js:
```diff
@@ -1231,41 +1340,44 @@ User = (function () {
 		}
 	};
-	User.prototype.ban = function (noRecurse, userid) {
+	User.prototype.ban = function (noRecurse, userid, expiration) {
```
Incidentally, `expiration` is a weird name for a variable. Maybe `expireTime`.

Zarel replied:
It should also go before `userid`, since `userid` is mostly internal and `expireTime` is external.

Zarel commented on users.js:
```diff
@@ -1273,10 +1385,11 @@ User = (function () {
 		}
-		if (this.autoconfirmed) lockedUsers[this.autoconfirmed] = this.userid;
-		if (this.authenticated) lockedUsers[this.userid] = this.userid;
+		var entry = [this.userid, expiration];
```
Why `this.userid` and not `userid`?

sirDonovan replied:
That's how it is in the current code, but now that I think about it, it would make more sense to track the <code>userid</code> like we do for bans.

Zarel commented on users.js:
```diff
@@ -156,17 +254,17 @@ function unlock(name, unlocked, noRecurse) {
 	for (var ip in lockedIps) {
-		if (userips && (ip in user.ips) && Users.lockedIps[ip] !== userid) {
-			unlocked = unlock(Users.lockedIps[ip], unlocked, true); // avoid infinite recursion
+		if (userips && (ip in user.ips) && Users.lockedIps[ip][0] !== userid) {
```
`Users.lockedIps[ip][0]` isn't this a bit dangerous if it might not be an array?

sirDonovan replied:
That's true, updated.

Zarel replied:
(This is what I meant when I said "Or your other code is very unsafe", btw)

sirDonovan commented:
Are there any other changes that need to be made?

ascriptmaster commented:
We were discussing a week-ban feature as well for serious cases, as an intermediate between the standard ban and permaban, so maybe you could toss that in somewhere.

sirDonovan commented:
I think that belongs in a separate pull request with the other timed punishments, which I'll make after this is implemented.

Zarel commented:
Yeah, just hold on, I'll have time for this eventually.

ascriptmaster commented:
bump

kotarou3 commented:
bump

Zarel commented:
This needs to be rebased, which I'll do since I have a few other changes I need to make to it, too.

Zarel commented:
Unlike the other long-running pullreqs, I don't need to optimize main process RAM usage too much for this, so I should be able to get to this soon.

Joimer commented:
"We can’t automatically merge this pull request. "
:(

Zarel commented:
Just as well; I needed to manually merge it either way.

Joimer commented:
*knock* *knock*

Zarel commented:
is still on my list

Zarel commented:
I've recently started optimizing, too! I told you about that.

Joimer commented:
Yeah, sorry, I thought this was ready for manual merge.

Joimer commented:
:^)

Zarel commented:
Still on my to-do list, I haven't forgotten about it.

ascriptmaster commented:
one day... one day

Zarel commented:
I check Issues all the time, so no, I haven't forgotten this.
